pkg/manifestclient/read_roundtripper.go

```diff
@@ -128,10 +140,94 @@ func isServerGroupResourceDiscovery(path string) bool {
 	if path == "/api/v1" {
 		return true
 	}
+	if path == "/api" {
+		return true
+	}
 
 	parts := strings.Split(path, "/")
 	if len(parts) != 4 {
 		return false
 	}
 	return parts[0] == "" && parts[1] == "apis"
 }
+
+//go:embed default-discovery
+var defaultDiscovery embed.FS
+
+func (mrt *manifestRoundTripper) getKindForResource(gvr schema.GroupVersionResource) (kindData, error) {
+	mrt.lock.RLock()
+	kindForGVR, ok := mrt.kindForResource[gvr]
+	if ok {
+		defer mrt.lock.RUnlock()
+		return kindForGVR, kindForGVR.err
+	}
+	mrt.lock.RUnlock()
+
+	mrt.lock.Lock()
+	defer mrt.lock.Unlock()
+
+	kindForGVR, ok = mrt.kindForResource[gvr]
+	if ok {
+		return kindForGVR, kindForGVR.err
+	}
+
+	discoveryPath := "/apis"
+	if len(gvr.Group) == 0 {
+		discoveryPath = "/api"
+	}
+	discoveryBytes, err := mrt.getGroupResourceDiscovery(&apirequest.RequestInfo{Path: discoveryPath})
+	if err != nil {
+		kindForGVR.err = fmt.Errorf("error reading discovery: %w", err)
+		mrt.kindForResource[gvr] = kindForGVR
+		return kindForGVR, kindForGVR.err
+	}
+
+	discoveryInfo := &apidiscoveryv2.APIGroupDiscoveryList{}
+	if err := json.Unmarshal(discoveryBytes, discoveryInfo); err != nil {
+		kindForGVR.err = fmt.Errorf("error unmarshalling discovery: %w", err)
+		mrt.kindForResource[gvr] = kindForGVR
+		return kindForGVR, kindForGVR.err
+	}
+
+	kindForGVR.err = fmt.Errorf("did not find kind for %v\n", gvr)
+	for _, groupInfo := range discoveryInfo.Items {
+		if groupInfo.Name != gvr.Group {
+			continue
+		}
+		for _, versionInfo := range groupInfo.Versions {
+			if versionInfo.Version != gvr.Version {
+				continue
+			}
+			for _, resourceInfo := range versionInfo.Resources {
+				if resourceInfo.Resource != gvr.Resource {
+					continue
+				}
+				if resourceInfo.ResponseKind == nil {
+					continue
+				}
+				kindForGVR.kind = schema.GroupVersionKind{
+					Group:   gvr.Group,
+					Version: gvr.Version,
+					Kind:    resourceInfo.ResponseKind.Kind,
+				}
+				if len(resourceInfo.ResponseKind.Group) > 0 {
+					kindForGVR.kind.Group = resourceInfo.ResponseKind.Group
+				}
+				if len(resourceInfo.ResponseKind.Version) > 0 {
+					kindForGVR.kind.Version = resourceInfo.ResponseKind.Version
+				}
+				kindForGVR.listKind = schema.GroupVersionKind{
+					Group:   kindForGVR.kind.Group,
+					Version: kindForGVR.kind.Version,
+					Kind:    resourceInfo.ResponseKind.Kind + "List",
```

Comment:
I think this won't work for CRDs.

Reply:
> I think this won't work for CRDs.

Technically not correct.  Practically the best option we have.  Turns out we don't include ListKind in discovery.